internal/model/ruleset.go

```diff
@@ -0,0 +1,139 @@
+package model
+
+import "github.com/hashicorp/terraform-plugin-framework/types"
+
+type AggregationRuleSet []AggregationRule
+
+func (a AggregationRuleSet) ToTF(segment types.String) RuleSetTF {
+	output := make([]RuleSetRuleTF, len(a))
+	for i, rule := range a {
+		output[i] = rule.ToRuleSetRuleTF()
+	}
+
+	return RuleSetTF{
+		Segment: segment,
+		Rules:   output,
+	}
+}
+
+type RuleSetTF struct {
+	Segment types.String    `tfsdk:"segment"`
+	Rules   []RuleSetRuleTF `tfsdk:"rules"`
+}
+
+func (r RuleSetTF) ToAPIReq() []AggregationRule {
+	output := make([]AggregationRule, len(r.Rules))
+	for i, rule := range r.Rules {
+		output[i] = rule.ToAPIReq()
+	}
+	return output
+}
+
+// RuleSetRule is a subset of RuleTF that is used in the RuleSetTF struct
+// This is necessary because the tfsdk doesn't support embedding structs.
+type RuleSetRuleTF struct {
+	Metric    types.String `tfsdk:"metric"`
+	MatchType types.String `tfsdk:"match_type"`
+
+	Drop       types.Bool     `tfsdk:"drop"`
+	KeepLabels []types.String `tfsdk:"keep_labels"`
+	DropLabels []types.String `tfsdk:"drop_labels"`
+
+	Aggregations []types.String `tfsdk:"aggregations"`
+
+	AggregationInterval types.String `tfsdk:"aggregation_interval"`
+	AggregationDelay    types.String `tfsdk:"aggregation_delay"`
+}
+
+func (r AggregationRule) IsExactMatch() bool {
+	return r.MatchType == "" || r.MatchType == "exact"
+}
+
+func (r RuleSetRuleTF) ToAPIReq() AggregationRule {
+	return AggregationRule{
+		Metric:    r.Metric.ValueString(),
+		MatchType: r.MatchType.ValueString(),
+
+		Drop:       r.Drop.ValueBool(),
+		KeepLabels: toStringSlice(r.KeepLabels),
+		DropLabels: toStringSlice(r.DropLabels),
+
+		Aggregations: toStringSlice(r.Aggregations),
+
+		AggregationInterval: r.AggregationInterval.ValueString(),
+		AggregationDelay:    r.AggregationDelay.ValueString(),
+
+		ManagedBy: managedByTF,
+	}
+}
+
+// AlignUpstreamWithState reorders the rules from upstream to match the state.
+// This must mantain the semantic equality of the rules, but may reorder them as
+// long as all non-exact rules are present in the same order as they were
+// originally.
+func AlignUpstreamWithState(state AggregationRuleSet, upstream AggregationRuleSet) AggregationRuleSet {
+	// If the two rulesets are semantically different, then we can't reorder them
+	// without changing the meaning of the rules.
+	if !semanticallyEqualOrdering(state, upstream) {
+		return upstream
+	}
+
+	output := make(AggregationRuleSet, 0, max(len(state), len(upstream)))
+
+	// we know there can be no duplicates by metric name, so we can use a map
+	upstreamMap := make(map[string]AggregationRule, len(upstream))
+	for _, rule := range upstream {
+		upstreamMap[rule.Metric] = rule
```

Comment:
how would this behave if there is an exact matcher rule for a metric and a prefix matcher rule which matches on a prefix that is exactly equal as that metric name? 

I don't think this is an impossible use-case, even if it's unlikely, but there's always that one user.

Reply:
That's technically prevented by our API right now: https://github.com/grafana/backend-enterprise/blob/2929599609d4cbd7fdba0cd0b34456d994f74399/pkg/aggregation/rules/validation.go#L132-L134

Reply:
looks like our validation would fail in this case: https://github.com/grafana/backend-enterprise/blob/cdebdf75d52d8dff83093229815e262c0322cad8/pkg/aggregation/rules/validation.go#L132

not sure if this should be possible though :thinking:, anyway let's just keep it invalid for now.